src/uwtools/config.py

```diff
@@ -77,57 +78,88 @@ def replace_templates(self):
+    '''Class for YAML loader and support for in place value subsitution and include'''
 
     def __init__(self, config_path):
-
         ''' Load the file and update the dictionary '''
-
         super().__init__(config_path)
-
+        self.config_obj = None
         self.update(self._load())
 
     def _load(self):
-        ''' Load the user-provided YAML config file path into a dict
-        object. '''
-
+        ''' Load the user-provided YAML config file path into a dict object.'''
         with open(self.config_path, 'r', encoding="utf-8") as file_name:
-            cfg = yaml.load(file_name, Loader=yaml.SafeLoader)
-        return cfg
+            config = yaml.load(file_name, Loader=yaml.Loader)
+            self.config_obj = config
+            self.update(config)
+        return config
+
+    # This is the overloaded method for UserDict to realize dot notation resolution
+    # it has moved this back to a Config class which is now a Abstract Class
+    def __getattr__(self, item):
+        if item in self:
+            return self.__dict__["data"][item]
+            #return self[item]
+        raise AttributeError(f"'{type(self)}' object has no attribute '{item}'")
+
+    def parse_include(self,config_file=None,data=None,from_environment=None,replace_realtime=None):
+        ''' Sample code needed to implement the !INCLUDE tag '''
+        if config_file is not None:
+            with open(config_file,encoding='utf-8') as _file:
+                config = yaml.load(_file, Loader=yaml.Loader)
+            self.config_path = config_file
+        else:
+            config = data
+        if from_environment:
+            config = Template.substitute_structure_from_environment(config)
+        config = Template.substitute_structure(config,TemplateConstants.
+                                               DOLLAR_PARENTHESES,self.get)
+        if replace_realtime:
+            config = Template.substitute_structure(config,
+                     TemplateConstants.DOUBLE_CURLY_BRACES,self.get)
+        config = Template.substitute_with_dependencies(config,config,
+                 TemplateConstants.DOLLAR_PARENTHESES,shallow_precedence=False)
+        if config is not None:
+            self.update(config)
+        self.config_obj = config
+        return config
 
     def dump_file(self, output_path):
         ''' Write the dictionary to a YAML file '''
-
-        with open(output_path, 'w', encoding="utf-8") as file_name:
-            yaml.dump(self.data, file_name, sort_keys=False)
-
-
+        with open(output_path, 'w+', encoding="utf-8") as file_name:
+            if self.config_obj is not None:
+                yaml.dump(self.config_obj, file_name, sort_keys=False)
 class F90Config(Config):
 
     ''' Concrete class to handle Fortran namelist files. '''
 
-    def __init__(self, config_path):
-
+    def __init__(self, config_path=None,data=None):
         ''' Load the file and update the dictionary '''
```

Comment:
Yes I like this distinction and naming conventions makes more since .. adding update() to the configure classes going forward instead of using data=None.